core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java

```diff
@@ -741,7 +741,9 @@ public boolean append(Object kbase, long koff, int klen, Object vbase, long voff
         longArray.set(pos * 2 + 1, keyHashcode);
         isDefined = true;
 
-        if (numKeys >= growthThreshold && longArray.size() < MAX_CAPACITY) {
+        // We use two array entries per key, so the array size is twice the capacity.
+        // We should compare the current capacity of the array, instead of its size.
+        if (numKeys >= growthThreshold && longArray.size() / 2 < MAX_CAPACITY) {
           try {
             growAndRehash();
```

Comment:
Actually, I also think that we should false canGrowArray like: 

```scala
if (numKeys >= growthThreshold && longArray.size() / 2 >= MAX_CAPACITY) {
  canGrowArray = false;
}
```

So as we reach max capacity of the map, canGrowArray is set to false. We can fail next append and let the map spill and fallback to sort-based aggregation in HashAggregate. Thus we can prevent a similar forever-loop happens when we reach max capacity. 

cc @cloud-fan @felixcheung

Reply:
If we do this, we won't call `growAndRehash` here, is it expected?

Reply:
Oh, I was not meaning to replace current condition, but to add anoother check.

Reply:
let me think about it. if making sense, will submit another PR for it.

Reply:
@viirya I'm encountering the same problem that you describe here. When the map is close to `MAX_CAPACITY` and needs to grow, `numKeys >= growthThreshold && longArray.size() / 2 >= MAX_CAPACITY` is true. This prevents the map from resizing, but currently `canGrowArray` remains `true`. Therefore the map keeps accepting new keys and exceeds its growth threshold. This ultimately causes the query to fail in the UnsafeKVExternalSorter constructor.

It looks like you didn't submit a PR for this - is there a reason why not? If there's no problem with your suggested fix, I can submit a PR now.

Reply:
I submitted another PR targeting the case of reaching `MAX_CAPACITY`. It actually prevents we increase the number of keys to be `MAX_CAPACITY`.

Reply:
Thanks for the quick response! I saw that PR (https://github.com/apache/spark/pull/26914) but I don't think it solves the problem I'm encountering. That PR stops accepting new keys once we have reached `MAX_CAPACITY - 1` keys, but this is too late. By that time, we will have far exceeded the growth threshold. If we attempt to spill the map in this state, the UnsafeKVExternalSorter will not be able to reuse the long array for sorting, causing the query to fail.

Reply:
I think the problem I posted above, is when we reach `MAX_CAPACITY`, a forever-loop happens during calling lookup. The previous PR fixed it. Sounds like you are encountering another problem?

Reply:
>  If we attempt to spill the map in this state, the UnsafeKVExternalSorter will not be able to reuse the long array for sorting, causing the query to fail.

In `UnsafeKVExternalSorter`, it will check if the long array can be reused or not. Isn't? If it cannot be reused, a new pointer array will be created, no?

Reply:
> Sounds like you are encountering another problem?

You're right, it's not the same problem.

> In UnsafeKVExternalSorter, it will check if the long array can be reused or not. Isn't? If it cannot be reused, a new pointer array will be created, no?

Yes, but by this point the task has typically consumed all available memory, so the allocation of the new pointer array is likely to fail.

Reply:
I filed [SPARK-32872](https://issues.apache.org/jira/browse/SPARK-32872) and submitted https://github.com/apache/spark/pull/29744 to fix this.